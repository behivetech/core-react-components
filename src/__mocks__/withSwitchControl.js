import {omit} from 'lodash';

export default (WrappedComponent) => function SwitchControlled(props) { 
    return <WrappedComponent {...omit(props, ['formState'])} />; 
};