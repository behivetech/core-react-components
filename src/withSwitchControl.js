// Vendor Libs
import React, {Component, createRef} from 'react';
import PropTypes from 'prop-types';
import {difference, omit, union} from 'lodash';
import classnames from 'classnames';
import {MDCFormField} from '@material/form-field';

// Components
import withFormControl from './withFormControl';

export default function (WrappedComponent) {
    class SwitchControlled extends Component {
        constructor(props) {
            super(props);
            this.state = {
                checked: (props.checked || props.value === props.formState.getFieldValue(props.name)),
            };
            this.formFieldRef = createRef();
            this.handleChange = this.handleChange.bind(this);
        }

        componentDidMount() {
            const formField = new MDCFormField(this.formFieldRef.current);
            const {checked, formState, name, value} = this.props;

            formField.input = this.formFieldRef.current.getElementsByTagName('input')[0];

            if (checked) {
                formState.setFieldValue(name, value);
            }
        }

        handleChange(event) {
            const checked = event.target.checked;
            const {
                formState,
                name,
                onChange,
                value,
            } = this.props;
            const stateName = name.replace('[]', '');
            let returnValue = value;

            if (!checked) {
                returnValue = (value === true) ? false : '';
            }

            if(name !== stateName) {
                const stateValue = formState.getFieldValue(stateName) || [];

                returnValue = (checked)
                    ? union([...stateValue], [value])
                    : difference([...stateValue], [value]);
            }

            this.setState({checked});
            formState.setFieldValue(stateName, returnValue);
            onChange(event);
        }

        getClass() {
            return classnames('mdc-form-field', this.props.className);
        }

        render() {
            const {label, id, name} = this.props;

            return (
                <div className={this.getClass()} ref={this.formFieldRef}>
                    <WrappedComponent
                        {...omit(this.props, ['formState'])}
                        checked={this.state.checked}
                        onChange={this.handleChange}
                    />
                    {
                        (label)
                            ? <label htmlFor={id || name}>{label}</label>
                            : null
                    }
                </div>
            );
        }
    }

    SwitchControlled.propTypes = {
        /** ability to add additional className to the component for addistional styling from parent */
        className: PropTypes.string,
        /** checked attribute for the input field */
        checked: PropTypes.bool,
        formState: PropTypes.shape({
            /** Gets the field value from the FormControlled component */
            getFieldValue: PropTypes.func.isRequired,
            /** Sets the field value in the FormControlled to be a part of the JSON that will be pushed. */
            setFieldValue: PropTypes.func.isRequired,
        }),
        /** id attribute for the input field */
        id: PropTypes.string,
        /** the label to be shown with the checkbox */
        label: PropTypes.string,
        /** Name of the field. Required to be used as a key as part of the JSON that will be pushed */
        name: PropTypes.string.isRequired,
        /** Callback function for additional functionality when the component has a change */
        onChange: PropTypes.func,
        /** Value of the component when it is checked. Defaults to true/false if nothing is set */
        value: PropTypes.oneOfType([
            PropTypes.bool,
            PropTypes.string,
            PropTypes.number,
        ]),
    };

    SwitchControlled.defaultProps = {
        onChange: () => null,
        value: true,
    };

    return withFormControl(SwitchControlled);
}

