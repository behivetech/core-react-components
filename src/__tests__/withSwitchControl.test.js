import withSwitchControl from '../withSwitchControl';
jest.mock('../withFormControl');

describe('withSwitchControl', () => {
    const MockComponentWrapped = (props) => <div />;
    const MockComponentControlled = withSwitchControl(MockComponentWrapped);
    // eslint-disable-next-line
    const MockComponent = (props) => <MockComponentControlled {...props} />;
    const getEvent = (checked = false) => {
        return {
            target: {checked},
        };
    };

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render a component and correct props are set', () => {
        const wrapper = mount(<MockComponent name="mockName" className="mock-class" />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('MockComponentWrapped').props().checked).toBe(false);
        expect(wrapper.find('MockComponentWrapped').props().value).toBe(true);
    });

    it('should set a value if checked is true', () => {
        const wrapperInstance = shallow(
            <MockComponent name="mockName" className="mock-class" checked />
        ).dive().dive().instance();

        expect(wrapperInstance.props.formState.setFieldValue).toHaveBeenCalledWith('mockName', true);
    });

    it('should set states when checkbox is checked and onChange prop is called', () => {
        const mockOnChange = jest.fn();
        const mockEvent = getEvent(true);
        const wrapperInstance = shallow(
            <MockComponent name="mockName" className="mock-class" checked onChange={mockOnChange} />
        ).dive().dive().instance();

        wrapperInstance.handleChange(mockEvent);
        expect(wrapperInstance.state.checked).toBe(true);
        expect(wrapperInstance.props.formState.setFieldValue).toHaveBeenCalledWith('mockName', true);
        expect(mockOnChange).toHaveBeenCalledWith(mockEvent);
    });

    it('should set states when checkbox is not checked checked', () => {
        const wrapperInstance = shallow(
            <MockComponent name="mockName" className="mock-class" checked />
        ).dive().dive().instance();

        wrapperInstance.handleChange(getEvent());
        expect(wrapperInstance.state.checked).toBe(false);
        expect(wrapperInstance.props.formState.setFieldValue).toHaveBeenCalledWith('mockName', false);
    });

    it('should set states for arrays', () => {
        const wrapperInstance = shallow(
            <MockComponent name="mockName[]" className="mock-class" checked value="mock value" />
        ).dive().dive().instance();

        wrapperInstance.props.formState.getFieldValue.mockReturnValue(undefined);
        wrapperInstance.handleChange(getEvent(true));
        expect(wrapperInstance.props.formState.setFieldValue).toHaveBeenCalledWith('mockName', ['mock value']);
        wrapperInstance.props.formState.getFieldValue.mockReturnValue(['mock1 value']);
        wrapperInstance.handleChange(getEvent(true));
        expect(wrapperInstance.props.formState.setFieldValue).toHaveBeenCalledWith('mockName', ['mock1 value',  'mock value']);
    });

    it('should unset states for arrays', () => {
        const wrapperInstance = shallow(
            <MockComponent name="mockName[]" className="mock-class" checked value="mock value" />
        ).dive().dive().instance();

        wrapperInstance.props.formState.getFieldValue.mockReturnValue(['mock value']);
        wrapperInstance.handleChange(getEvent(false));
        expect(wrapperInstance.props.formState.setFieldValue).toHaveBeenCalledWith('mockName', []);
        wrapperInstance.props.formState.getFieldValue.mockReturnValue(['mock1 value', 'mock value']);
        wrapperInstance.handleChange(getEvent(false));
        expect(wrapperInstance.props.formState.setFieldValue).toHaveBeenCalledWith('mockName', ['mock1 value']);
    });
});