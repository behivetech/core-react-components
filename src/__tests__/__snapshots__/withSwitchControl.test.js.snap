// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withSwitchControl should render a component and correct props are set 1`] = `
<MockComponent
  className="mock-class"
  name="mockName"
>
  <FormFieldControlled
    className="mock-class"
    name="mockName"
  >
    <SwitchControlled
      className="mock-class"
      formState={
        Object {
          "formDisabled": false,
          "getFieldError": [MockFunction],
          "getFieldValue": [MockFunction] {
            "calls": Array [
              Array [
                "mockName",
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          },
          "initialSubmit": false,
          "setFieldValidators": [MockFunction],
          "setFieldValue": [MockFunction],
          "setFieldValueDebounced": [MockFunction],
          "shouldDisable": [MockFunction],
        }
      }
      name="mockName"
      onChange={[Function]}
      value={true}
    >
      <div
        className="mdc-form-field mock-class"
      >
        <MockComponentWrapped
          checked={false}
          className="mock-class"
          name="mockName"
          onChange={[Function]}
          value={true}
        >
          <div />
        </MockComponentWrapped>
      </div>
    </SwitchControlled>
  </FormFieldControlled>
</MockComponent>
`;
